org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/wizard/template/AbstractProjectTemplate.java

```diff
@@ -0,0 +1,122 @@
+ * present the template to the user and a list or {@link ProjectFile} elements to be generated by the wizard when the template is selected.
+ * The {@link ProjectFile}'s can contain references to {@link ProjectVariable}'s to allow the user to configure the generated code.
+ * 
+ * Instances of this class are generated from classes annotated with {@link ProjectTemplate}.
+ * 
+ * @author Arne Deutsch - Initial contribution and API
+ * @since 2.14
+ */
+public abstract class AbstractProjectTemplate {
+
+	private static final Logger logger = Logger.getLogger(AbstractProjectTemplate.class);
+
+	private List<ProjectVariable> variables;
+
+	protected final StringProjectVariable text(String name, String defaultValue, String description) {
+		return new StringProjectVariable(name, defaultValue, description);
+	}
+
+	protected final BooleanProjectVariable check(String name, boolean defaultValue, String description) {
+		return new BooleanProjectVariable(name, defaultValue, description);
+	}
+
+	protected final StringSelectionProjectVariable combo(String name, String[] possibleValues, String description) {
+		return new StringSelectionProjectVariable(name, possibleValues, description);
+	}
+
+	/**
+	 * @return The label read from the {@link ProjectTemplate} annotation.
+	 */
+	public final String getLabel() {
+		return getProjectTemplateAnnotation().label();
+	}
+
+	/**
+	 * @return The icon read from the {@link ProjectTemplate} annotation.
+	 */
+	public final String getIcon() {
+		return getProjectTemplateAnnotation().icon();
+	}
+
+	/**
+	 * @return The description read from the {@link ProjectTemplate} annotation.
+	 */
+	public final String getDescription() {
+		return getProjectTemplateAnnotation().description();
+	}
+
+	/**
+	 * @return The variables read reflectively from this class.
+	 */
+	public final List<ProjectVariable> getVariables() {
+		if (variables == null) {
+			// lazy initialization because initialization in constructor would initialize lists with null's because field initialization
+			// comes after constructor initialization of super class
+			variables = new ArrayList<>();
+			Class<? extends Object> projectTemplateClass = getClass();
+			for (Field field : projectTemplateClass.getDeclaredFields()) {
+				if (ProjectVariable.class.isAssignableFrom(field.getType())) {
+					try {
+						field.setAccessible(true);
+						variables.add((ProjectVariable) field.get(this));
+					} catch (IllegalArgumentException | IllegalAccessException e) {
+						logger.error("Can not create project variable for '" + field.getName() + "'", e);
+					}
+				}
+			}
+			variables = Collections.unmodifiableList(variables);
+		}
+		return variables;
+	}
+
+	/**
+	 * Subclasses have to override. Generate all the files to be created when the wizard is finished.
+	 */
+	abstract public void generateFiles(IProjectFileGenerator generator);
```

Comment:
minor: canonical order of modifiers would be `public abstract`

Reply:
Will change that.